### PR TITLE
HADOOP-19090. Use protobuf-java 3.23.4.

### PR DESCRIPTION
**Pull request by pjfanning**
### Description of PR

DRAFT - See HADOOP-19090

The hadoop-thirdparty change is not merged yet so I am temporarily testing with a snapshot that I published with my groupId.

### How was this patch tested?

Local build

### For code changes:

- [x] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

File: hadoop-project/pom.xml [hunk 1]
```diff
@@ -93,11 +93,11 @@
     <!-- Protobuf scope in other modules which explicitly import the libarary -->
     <transient.protobuf2.scope>${common.protobuf2.scope}</transient.protobuf2.scope>
     <!-- ProtocolBuffer version, actually used in Hadoop -->
-    <hadoop.protobuf.version>3.21.12</hadoop.protobuf.version>
+    <hadoop.protobuf.version>3.23.4</hadoop.protobuf.version>
```
Comment (steveloughran) [inline hadoop-project/pom.xml:96]: I'd prefer editing hadoop-thirdparty.version

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  11m 44s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  1s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  15m 14s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  33m 51s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 40s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  compile  |  16m 15s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  mvnsite  |  20m 56s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   8m 36s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   7m 40s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  | 152m 50s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 52s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  30m 57s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m 55s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javac  |  16m 55s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  16m  2s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |  16m  2s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |  16m 13s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   8m 28s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   7m 51s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  66m 23s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 737m 39s | [/patch-unit-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/1/artifact/out/patch-unit-root.txt) |  root in the patch failed.  |
| +1 :green_heart: |  asflicense  |   1m 35s |  |  The patch does not generate ASF License warnings.  |
|  |   | 1039m 44s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.namenode.TestSecureNameNode |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.44 ServerAPI=1.44 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6593 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint |
| uname | Linux bc53b5a08ee5 5.15.0-94-generic #104-Ubuntu SMP Tue Jan 9 15:25:40 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 53109b7dab02e7cc8f756231ae1b2f3f81d91799 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/1/testReport/ |
| Max. process+thread count | 4799 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api . U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/1/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-project/pom.xml (hunk 1 above)
Comment (pjfanning) [reply hadoop-project/pom.xml:96]: this value is the one used by the protoc compilation and the build fails if I don't upgrade it

**Comment by pjfanning** [comment]
The 1 broken test seems unrelated:

```
[ERROR] org.apache.hadoop.hdfs.server.namenode.TestSecureNameNode  Time elapsed: 1.417 s  <<< ERROR!
org.apache.kerby.kerberos.kerb.KrbException: Failed to start DefaultKrbServer. Address already in use (Bind failed)
```

File: hadoop-project/pom.xml (hunk 1 above)
Comment (ayushtkn) [reply hadoop-project/pom.xml:96]: this is correct, you need to update the version for the protoc generation, thirdparty protobuf isn't used for generating the protoc classes. We generate the classes & then do a replacement in those genrated classes to use the shaded protobuf

https://github.com/apache/hadoop/blob/10ab8abccd7e086d1d47eb7d706c8d5c42052172/hadoop-project/pom.xml#L2269-L2292

**Review by ayushtkn — COMMENTED** [review]
the change is merged to hadoop-thirdparty & we have a automated build which publishes the snapshot version of that, can we use that, if the build is green we can merge this to get the tests fixed

**Comment by pjfanning** [comment]
Switched PR to use the Apache snapshot

**Comment by Hexiaoqiao** [comment]
Thanks @pjfanning . LGTM. Let's wait what CI will say.
Try link to https://github.com/apache/hadoop-thirdparty/pull/34

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  11m 57s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 38s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  32m  8s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  17m 29s |  |  trunk passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  compile  |  15m 56s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  mvnsite  |  21m 29s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   8m 36s |  |  trunk passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   7m 37s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  | 151m 56s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 53s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  32m 13s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  18m 16s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javac  |  18m 16s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  17m 14s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |  17m 14s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |  17m 50s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   8m 52s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   8m  1s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  72m 10s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 731m 59s | [/patch-unit-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/2/artifact/out/patch-unit-root.txt) |  root in the patch failed.  |
| +1 :green_heart: |  asflicense  |   1m 38s |  |  The patch does not generate ASF License warnings.  |
|  |   | 1043m  4s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.yarn.server.resourcemanager.reservation.TestCapacityOverTimePolicy |
|   | hadoop.hdfs.tools.TestDFSAdmin |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.44 ServerAPI=1.44 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6593 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint |
| uname | Linux d622216aa576 5.15.0-91-generic #101-Ubuntu SMP Tue Nov 14 13:30:08 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 9a7e4e611dfb5fc6bca1c5a7f6390805ec899334 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/2/testReport/ |
| Max. process+thread count | 4498 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api . U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/2/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: pom.xml [hunk 2]
```diff
@@ -57,6 +57,13 @@ xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/x
       <name>${distMgmtSnapshotsName}</name>
       <url>${distMgmtSnapshotsUrl}</url>
     </repository>
+    <repository>
+      <id>Apache Nexus Snapshots</id>
+      <url>https://repository.apache.org/content/groups/snapshots/</url>
+      <snapshots>
+        <enabled>true</enabled>
+      </snapshots>
+    </repository>
```
Comment (ayushtkn) [inline pom.xml]: do we need this? We have a repo defined for snapshots already
```
    <distMgmtSnapshotsUrl>https://repository.apache.org/content/repositories/snapshots</distMgmtSnapshotsUrl>
```

Comment (pjfanning) [reply pom.xml]: removed

**Review by ayushtkn — APPROVED** [review]
LGTM

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   6m 38s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 16s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  19m 59s |  |  trunk passed  |
| +1 :green_heart: |  compile  |   9m  2s |  |  trunk passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  compile  |   8m  4s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  mvnsite  |  13m  7s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   4m 49s |  |  trunk passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   4m 58s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  93m 28s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 39s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  18m 30s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m 41s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javac  |   8m 41s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   8m  0s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |   8m  0s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |   8m 20s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   4m 44s |  |  the patch passed with JDK Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1  |
| +1 :green_heart: |  javadoc  |   5m  7s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  38m 21s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 548m 49s | [/patch-unit-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/3/artifact/out/patch-unit-root.txt) |  root in the patch failed.  |
| +0 :ok: |  asflicense  |   0m 34s |  |  ASF License check generated no output?  |
|  |   | 726m 29s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.yarn.server.timelineservice.storage.flow.TestHBaseStorageFlowActivity |
|   | hadoop.yarn.server.timelineservice.storage.TestHBaseTimelineStorageSchema |
|   | hadoop.hdfs.tools.TestDFSAdmin |
|   | hadoop.hdfs.TestErasureCodingMultipleRacks |
|   | hadoop.tools.TestDistCpWithRawXAttrs |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.44 ServerAPI=1.44 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6593 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint |
| uname | Linux 279b9de865a7 5.15.0-94-generic #104-Ubuntu SMP Tue Jan 9 15:25:40 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 2035ce9e797acf655b1eb576b2164353a69d8bb3 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.22+7-post-Ubuntu-0ubuntu220.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/3/testReport/ |
| Max. process+thread count | 4428 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-yarn-project/hadoop-yarn/hadoop-yarn-api . U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/3/console |
| versions | git=2.25.1 maven=3.6.3 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by pjfanning** [comment]
Some of the test issues relate to `java.lang.OutOfMemoryError: unable to create new native thread`

```
[ERROR] Tests run: 10, Failures: 0, Errors: 5, Skipped: 0, Time elapsed: 418.936 s <<< FAILURE! - in org.apache.hadoop.tools.TestDistCpWithRawXAttrs
[ERROR] testUseIterator(org.apache.hadoop.tools.TestDistCpWithRawXAttrs)  Time elapsed: 155.529 s  <<< ERROR!
java.io.IOException: java.lang.OutOfMemoryError: unable to create new native thread
	at org.apache.hadoop.hdfs.ExceptionLastSeen.set(ExceptionLastSeen.java:45)
	at org.apache.hadoop.hdfs.DataStreamer.run(DataStreamer.java:861)
Caused by: java.lang.OutOfMemoryError: unable to create new native thread
	at java.lang.Thread.start0(Native Method)
	at java.lang.Thread.start(Thread.java:719)
	at org.apache.hadoop.hdfs.DataStreamer.initDataStreaming(DataStreamer.java:672)
	at org.apache.hadoop.hdfs.DataStreamer.run(DataStreamer.java:752)
```

**Review by Hexiaoqiao — APPROVED** [review]
+1 from my side.

**Comment by steveloughran** [comment]
we ready to get this into branch-3.4?

**Comment by pjfanning** [comment]
@steveloughran this PR depends on a snapshot version of hadoop-thirdparty. Can we do a release of hadoop-thirdparty and then backport the uptake of the future hadoop-thirdparty release to the 3.4 branch?

**Comment by steveloughran** [comment]
are we up to date with other dependencies for a new thirdparty release? we should really bump them all up and then rebuild it. 
this isn't targeting 3.4.0 so not a blocker there

**Comment by pjfanning** [comment]
> are we up to date with other dependencies for a new thirdparty release? we should really bump them all up and then rebuild it. this isn't targeting 3.4.0 so not a blocker there

Guava has a slightly newer release 33.0.0-jre. Is this upgrade needed though?

**Comment by zhangshuyan0** [comment]
It seems that `hadoop-shaded-protobuf_3_23` has not been deployed to the maven central repository yet. I cannot compile trunk with the latest code now. @pjfanning

**Comment by pjfanning** [comment]
@zhangshuyan0 it is in the Apache Snapshot repo. See the version number: 1.3.0-SNAPHOT.

See https://repository.apache.org/content/groups/snapshots/org/apache/hadoop/thirdparty/hadoop-shaded-protobuf_3_23/

See https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6593/3

**Comment by zhangshuyan0** [comment]
@pjfanning Thanks for your reply. The reason I couldn't compile successfully was due to my local network problem.

**Pull request merged**